### DISCUSSION TITLE
Example of healthy loss functions in tacotron2 multiband melgan

### DISCUSSION THREAD

**Discussion by JRMeyer** [coqui-ai/TTS#251] [Archived]
<USER>>>> rui<br/> <DATE> [August 8, 2020, 8:44pm]<br/> <br/> Hi,<br/> <br/> I am testing the approach with Tacotron2 + Multiband Melgan in the dev<br/> branch (commit 8c07ae7). *slash* <br/> I was wondering if anyone has an example of what a healthy loss function<br/> looks like when training the vocoder.<br/> <br/> In my case, (with a smallish custom dataset) I observed some weird<br/> behavior the moment we introduce the discriminator<br/> (`steps_to_start_discriminator`). *slash* <br/> Before this point the generator losses were trending down (as expected).<br/> After this point they somewhat blew up.<br/> <br/> Is this expected? Just a matter of training for longer? Maybe setting a<br/> larger `steps_to_start_discriminator` ?<br/> <br/> 472<br/> <br/> Thanks!<br/> <br/> [This is an archived TTS discussion thread from discourse.mozilla.org/t/example-of-healthy-loss-functions-in-tacotron2-multiband-melgan]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> dkreutz<br/> <DATE> [August 9, 2020, 9:08am]<br/> <br/> Recently there were some discussions here on discriminator training, to<br/> my understanding there are two options you can try: a) start<br/> discriminator training right away from the beginning or b) let<br/> discriminator training kick in far later, e.g. step 500k (default is<br/> IIRC at 200k)<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> rui<br/> <DATE> [August 9, 2020, 9:48am]<br/> <br/> Thanks Dominik. I will try both mechanisms and see what works best in my<br/> dataset.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> nmstoker<br/> <DATE> [August 9, 2020, 5:07pm]<br/> <br/> I haven't trained often enough to get a good sense of what works best on<br/> my dataset or overall. I did have a number of false starts and then by<br/> perseverance I ended up with a single quite good result, but it may well<br/> be not the optimal approach.<br/> <br/> My good run was done with the discriminator kicking in at 600k<br/> had made on<br/> his blog). When it does so, there's a long period where it definitely<br/> sounds like the audio you hear in Tensorboard got worse, but in time you<br/> could very gradually start to hear that it was actually getting better<br/> than it had been earlier on (before the discriminator kicked in).<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> rui<br/> <DATE> [August 9, 2020, 5:51pm]<br/> <br/> . *slash* <br/> Maybe I just need let it train for a longer period of time.<br/> <br/> In any case, can you share a printscreen of the losses you got on your<br/> good run ? It would be useful to have an idea on the magnitude of the<br/> losses I should aim for in order to get good results.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> nmstoker<br/> <DATE> [August 9, 2020, 11:40pm]<br/> <br/> Hi Rui,<br/> <br/> Here you go. I did have some computer memory issues (since resolved in<br/> dev branch) during this time, so for much of it i was stopping and<br/> restarting training which tends to disrupt the charts for a bit.<br/> <br/> I am wary to say 'just train longer' because it takes a long time to<br/> train and I'd feel bad if it turned out then that you trained for ages<br/> and got poor results (ie wasting your time) but I did find that for me<br/> the quality improvement started being most noticeable around 800-900k<br/> with it starting to eclipse my best prior model without a vocoder around<br/> then and it got to exceed that best prior model as it went from 900k to<br/> 1m. There was a gradual noticable improvement before that 800k point but<br/> it was a little odd as in some ways it still sounded worse than when the<br/> discriminator kicked in but there was a more human sound creeping in<br/> even though the quality was set back. Anyway, when I get around to doing<br/> another run, I hope to gain more of a sense of whether this was 'normal'<br/> or not. Maybe if others chip in we'll get a picture more quickly too.<br/> <br/> Hope it goes well and I'd be interested to hear updates<br/> <br/> Kind regards, *slash* <br/> Neil<br/> <br/> 315<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> nmstoker<br/> <DATE> [August 9, 2020, 11:47pm]<br/> <br/> BTW: here's a demo of the results using the vocoder from the 1m point<br/> (by that time I hadn't completed the 1m-1.2m run that's the pink line<br/> above; to be honest I'm in two minds as to whether it improved much<br/> during that final phase up to 1.2m)<br/> <br/> [Archived Post]<br/>
